doc/test-framework/writing-e2e-tests.md

```diff
@@ -0,0 +1,236 @@
+exampleMemcached := &cachev1alpha1.Memcached{
+    TypeMeta: metav1.TypeMeta{
+        Kind:       "Memcached",
+        APIVersion: "cache.example.com/v1alpha1",
+    },
+    ObjectMeta: metav1.ObjectMeta{
+        Name:      "example-memcached",
+        Namespace: namespace,
+    },
+    Spec: cachev1alpha1.MemcachedSpec{
+        Size: 3,
+    },
+}
+err = f.DynamicClient.Create(goctx.TODO(), exampleMemcached)
+if err != nil {
+    return err
+}
+```
+
+Now we can check if the operator successfully worked. In the case of the memcached operator, it should have
+created a deployment called "example-memcached" with 3 replicas:
+
+```go
+// wait for example-memcached to reach 3 replicas
+err = e2eutil.WaitForDeployment(t, f.KubeClient, namespace, "example-memcached", 3, time.Second*5, time.Second*30)
+if err != nil {
+    return err
+}
+```
+
+We can also test that the deployment scales correctly when the CR is updated:
+
+```go
+err = f.DynamicClient.Get(goctx.TODO(), types.NamespacedName{Name: "example-memcached", Namespace: namespace}, exampleMemcached)
+if err != nil {
+    return err
+}
+exampleMemcached.Spec.Size = 4
+err = f.DynamicClient.Update(goctx.TODO(), exampleMemcached)
+if err != nil {
+    return err
+}
+
+// wait for example-memcached to reach 4 replicas
+err = e2eutil.WaitForDeployment(t, f.KubeClient, namespace, "example-memcached", 4, time.Second*5, time.Second*30)
+if err != nil {
+    return err
+}
+```
+
+Now that the test is complete, we are done with the function. Once the end of the function is reached, the TestCtx's cleanup
+functions will automatically be run since they were deferred when the TestCtx was created.
+
+## Manual Cleanup
+
+While the test framework provides utilities that allow the test to automatically be cleaned up when done,
+it is possible that an error in the test code could cause a panic, which would stop the test
+without running the deferred cleanup. To clean up manually, you should check what namespaces currently exist
+in your cluster. You can do this with `kubectl`:
+
+```shell
+$ kubectl get namespaces
+
+Example Output:
+NAME                                            STATUS    AGE
+default                                         Active    2h
+kube-public                                     Active    2h
+kube-system                                     Active    2h
+main-1534287036                                 Active    23s
+memcached-memcached-group-cluster-1534287037    Active    22s
+memcached-memcached-group-cluster2-1534287037   Active    22s
+```
+
+The names of the namespaces will be either start with `main` or with the name of the tests and the suffix will
+be a Unix timestamp (number of seconds since January 1, 1970 00:00 UTC). Kubectl can be used to delete these
+namespaces and the resources in those namespaces:
+
+```shell
+$ kubectl delete namespace main-153428703
+```
+
+The tests also create a CRD. The CRD can be removed quite easily. The simplest way is to use `kubectl` and
```

Comment:
Rephrase:
`The tests also create a CRD. The CRD can be removed quite easily. The simplest way is to use `kubectl` and
the file that was used to create the crd (by default `deploy/crd.yaml`):`
to:
`Clean up the CRD created by the tests using the CRD manifest deploy/crd.yaml:`